File: timer_control/timer_control/scripts/timer_control.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from interfaces.srv import TimerControl
import time


class TimedCommandPublisher(Node):
    def __init__(self):
        super().__init__('timed_command_publisher')
        self.declare_parameter('command_topic', '/cmd_vel')
        self.declare_parameter('default_duration', 5.0)
        self.declare_parameter('default_linear_speed', 0.5)
        self.declare_parameter('default_angular_speed', 0.0)

        self.command_topic = self.get_parameter('command_topic').value
        self.default_duration = self.get_parameter('default_duration').value
        self.default_linear_speed = self.get_parameter('default_linear_speed').value
        self.default_angular_speed = self.get_parameter('default_angular_speed').value

        self.command_pub = self.create_publisher(Twist, self.command_topic, 10) #Publisher des contrôles en vitesse Twist

        self.trigger_service = self.create_service(TimerControl, 'trigger_command', self.trigger_command_callback)

        self.is_active = False
        self.get_logger().info("TimedCommandPublisher ready. Call '/trigger_command' to start a command.")

    def trigger_command_callback(self, request, response):
        if self.is_active:
            response.success = False
            response.report = "A command is already active."
            self.get_logger().warn("Cannot start a new command. One is already active.")
            return response

        duration = request.duration if request.duration > 0 else self.default_duration
        linear_speed = request.linear_vx if request.linear_vx != 0 else self.default_linear_speed
        angular_speed = request.angular_wz if request.angular_wz != 0 else self.default_angular_speed

        self.send_command(duration, linear_speed, angular_speed)
        response.success = True
        response.message = "Command executed successfully."
        return response

    def send_command(self, duration, linear_speed, angular_speed):
        cmd = Twist()
        cmd.linear.x = linear_speed
        cmd.angular.z = angular_speed

        self.get_logger().info(f"Publishing command for {duration:.2f} seconds.")
        self.is_active = True
        start_time = time.time()

        while time.time() - start_time < duration:
            self.command_pub.publish(cmd)
            time.sleep(0.1)  # 10 Hz

        self.stop_command()

    def stop_command(self):
        stop_cmd = Twist()
        stop_cmd.linear.x = 0.0
        stop_cmd.angular.z = 0.0

        self.command_pub.publish(stop_cmd)
        self.get_logger().info("Command stopped.")
        self.is_active = False

def main(args=None):
    rclpy.init(args=args)
    node = TimedCommandPublisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: tekbot_description/src/timer_control.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include "tekbot_description/srv/timer_control.hpp"
#include <nav_msgs/msg/odometry.hpp>

using std::placeholders::_1;
using std::placeholders::_2;

class TimedCommandPublisher : public rclcpp::Node {
public:
    TimedCommandPublisher() : Node("timed_command_publisher"), is_active_(false) {
        // Déclaration des paramètres
        declare_parameter<std::string>("command_topic", "/cmd_vel");
        declare_parameter<std::string>("odom_topic", "/odometry/ground_truth");
        declare_parameter<double>("default_duration", 5.0);
        declare_parameter<double>("default_linear_speed", 0.5);
        declare_parameter<double>("default_angular_speed", 0.0);

        // Récupération des paramètres
        command_topic_ = get_parameter("command_topic").as_string();
        odom_topic_ = get_parameter("odom_topic").as_string();
        default_duration_ = get_parameter("default_duration").as_double();
        default_linear_speed_ = get_parameter("default_linear_speed").as_double();
        default_angular_speed_ = get_parameter("default_angular_speed").as_double();

        // Initialisation des éditeurs et abonnements
        command_pub_ = create_publisher<geometry_msgs::msg::Twist>(command_topic_, 10);
        // odom_sub_ = create_subscription<nav_msgs::msg::Odometry>(
        //     odom_topic_, 10, std::bind(&TimedCommandPublisher::odomCallback, this, _1));
        
        // Service de commande
        trigger_service_ = create_service<tekbot_description::srv::TimerControl>(
            "trigger_command", std::bind(&TimedCommandPublisher::triggerCommandCallback, this, _1, _2));

        RCLCPP_INFO(get_logger(), "TimedCommandPublisher ready. Call '/trigger_command' to start a command.");
    }

private:
    // Publishers, subscribers et service
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr command_pub_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
    rclcpp::Service<tekbot_description::srv::TimerControl>::SharedPtr trigger_service_;

    std::string command_topic_, odom_topic_;
    double default_duration_, default_linear_speed_, default_angular_speed_;
    bool is_active_;

    void triggerCommandCallback(const tekbot_description::srv::TimerControl::Request::SharedPtr req,
                                tekbot_description::srv::TimerControl::Response::SharedPtr res) {
        if (is_active_) {
            res->success = false;
            res->message = "A command is already active.";
            RCLCPP_WARN(get_logger(), "Cannot start a new command. One is already active.");
            return;
        }

        double duration = req->duration > 0 ? req->duration : default_duration_;
        double linear_speed = req->linear_speed != 0 ? req->linear_speed : default_linear_speed_;
        double angular_speed = req->angular_speed != 0 ? req->angular_speed : default_angular_speed_;
        sendCommand(duration, linear_speed, angular_speed);
        res->success = true;
        res->message = "Command executed successfully.";
    }

    void sendCommand(double duration, double linear_speed, double angular_speed) {
        auto cmd = geometry_msgs::msg::Twist();
        cmd.linear.x = linear_speed;
        cmd.angular.z = angular_speed;

        auto rate = rclcpp::Rate(10); // Publier à 10 Hz
        RCLCPP_INFO(get_logger(), "Publishing command for %.2f seconds.", duration);

        auto start_time = now();
        is_active_ = true;

        while ((now() - start_time).seconds() < duration) {
            command_pub_->publish(cmd);
            rate.sleep();
        }

        stopCommand();
    }

    void stopCommand() {
        auto stop_cmd = geometry_msgs::msg::Twist();
        stop_cmd.linear.x = 0.0;
        stop_cmd.angular.z = 0.0;

        command_pub_->publish(stop_cmd);
        RCLCPP_INFO(get_logger(), "Command stopped.");
        is_active_ = false;
    }
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TimedCommandPublisher>());
    rclcpp::shutdown();
    return 0;
}
